math/src/fft/fft_inputs.rs

```diff
@@ -0,0 +1,92 @@
+use super::StarkField;
+use crate::FieldElement;
+
+// FFTINPUTS TRAIT
+// ================================================================================================
+
+/// Defines the shape of the input passed as input for FFT computation.
+pub trait FftInputs<B: StarkField> {
+    /// Returns the length of this fftinputs.
+    fn len(&self) -> usize;
+
+    /// Combines the result of smaller discrete fourier transforms into a larger DFT.
+    fn butterfly(&mut self, offset: usize, stride: usize);
+
+    /// Combines the result of smaller discrete fourier transforms multiplied with a
+    /// twiddle factor into a larger DFT.
+    fn butterfly_twiddle(&mut self, twiddle: B, offset: usize, stride: usize);
+
+    /// Swaps the element at index i with the element at index j.
+    fn swap(&mut self, i: usize, j: usize);
+
+    /// Helper method used to replace each element in fftinputs with its product with
+    /// `init_offset`, followed by `init_offset` getting updated by `offset_factor` right
+    /// after the element update. The mathematical formula is as follows:
+    ///
+    /// $$
+    /// elem_{i} = elem_{i} * init_offset_{i}
+    /// init_offset_{i+1} = init_offset_{i} * offset_factor
+    /// $$
+    ///
+    /// where elem is the element of the fftinput trait and i is the index of the element in
+    /// the fftindex.
+    fn shift_by_with_offset(&mut self, init_offset: B, offset_factor: B);
+
+    /// Helper method used to replace each element in fftinputs with its product with
+    /// `init_offset. The mathematical formula is as follows:
+    ///
+    /// $$
+    /// elem_{i} = elem_{i} * init_offset
+    /// $$
+    ///
+    /// where elem is the element of the fftinput trait and i is the index of the element in
+    /// the fftindex.
```

Comment:
I would change this to:
```
Multiplies every element in this input by `offset`. Specifically:

elem_{i} = elem_{i} * offset
```
Also, `init_offset` will not render correctly in the formula as underscores are used for subscripts. You can probably drop `$$` and it will be fine.

Reply:
done